File: 1408-find-the-smallest-divisor-given-a-threshold/find-the-smallest-divisor-given-a-threshold.py
class Solution(object):
    def smallestDivisor(self, nums, threshold):
        """
        :type nums: List[int]
        :type threshold: int
        :rtype: int
        """
        def solve(div):
            return sum((i + div - 1) // div for i in nums)
        low=1
        high=max(nums)
        res=0
        while(low<=high):
            mid=(low+high)//2
            ans=solve(mid)
            if(ans<=threshold):
                res=mid
                high=mid-1
            else:
                low=mid+1
        return res                

File: 1408-find-the-smallest-divisor-given-a-threshold/find-the-smallest-divisor-given-a-threshold.java
class Solution {
    public int smallestDivisor(int[] nums, int threshold) {
        int low=1;
        int high=Integer.MIN_VALUE;
        int ans=high;
        for(int i:nums){
            high=Math.max(high,i);
        }
        while(low<=high){
            int mid=(low+high)/2;
            int s=0;
            for(int i: nums){
                s+=(i+mid-1)/mid;
            }
            if(s<=threshold){
                ans=mid;
                high=mid-1;
            }
            else{
                low=mid+1;
            }
        }
        return ans;    
    }
}